File: algorithms/strings/trie.js
/**
 * Trie (Prefix Tree)
 * insert: O(m) | search: O(m) | startsWith: O(m) where m = word length
 *
 * Efficient for autocomplete, spell checking, and prefix-based search.
 */

class TrieNode {
  constructor() {
    this.children = {};
    this.isEndOfWord = false;
    this.count = 0; // Number of words with this prefix
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let node = this.root;
    for (const char of word) {
      if (!node.children[char]) {
        node.children[char] = new TrieNode();
      }
      node = node.children[char];
      node.count++;
    }
    node.isEndOfWord = true;
  }

  search(word) {
    const node = this._traverse(word);
    return node !== null && node.isEndOfWord;
  }

  startsWith(prefix) {
    return this._traverse(prefix) !== null;
  }

  // Get all words with a given prefix (autocomplete)
  autocomplete(prefix, maxResults = 10) {
    const node = this._traverse(prefix);
    if (!node) return [];
    const results = [];
    this._dfs(node, prefix, results, maxResults);
    return results;
  }

  _traverse(word) {
    let node = this.root;
    for (const char of word) {
      if (!node.children[char]) return null;
      node = node.children[char];
    }
    return node;
  }

  _dfs(node, prefix, results, maxResults) {
    if (results.length >= maxResults) return;
    if (node.isEndOfWord) results.push(prefix);
    for (const [char, child] of Object.entries(node.children)) {
      this._dfs(child, prefix + char, results, maxResults);
    }
  }
}

// --- Example ---
const trie = new Trie();
['apple', 'app', 'application', 'apt', 'bat', 'ball'].forEach(w => trie.insert(w));

console.log(trie.search('app'));       // true
console.log(trie.search('ap'));        // false
console.log(trie.startsWith('ap'));    // true
console.log(trie.autocomplete('app')); // ['app', 'apple', 'application']

module.exports = { Trie };

File: algorithms/trees/Trie.java
package algorithms.trees;

import java.util.*;

/**
 * Trie (Prefix Tree) — O(m) insert/search, m = word length
 */
public class Trie {

    private final TrieNode root;

    public Trie() { root = new TrieNode(); }

    public void insert(String word) {
        TrieNode curr = root;
        for (char c : word.toCharArray()) {
            curr.children.putIfAbsent(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.isEnd = true;
        curr.word = word;
    }

    public boolean search(String word) {
        TrieNode node = find(word);
        return node != null && node.isEnd;
    }

    public boolean startsWith(String prefix) {
        return find(prefix) != null;
    }

    private TrieNode find(String s) {
        TrieNode curr = root;
        for (char c : s.toCharArray()) {
            if (!curr.children.containsKey(c)) return null;
            curr = curr.children.get(c);
        }
        return curr;
    }

    // Autocomplete: all words with given prefix
    public List<String> autocomplete(String prefix) {
        List<String> result = new ArrayList<>();
        TrieNode node = find(prefix);
        if (node != null) dfs(node, result);
        return result;
    }

    private void dfs(TrieNode node, List<String> result) {
        if (node.isEnd) result.add(node.word);
        for (TrieNode child : node.children.values()) dfs(child, result);
    }

    // Word Search II — find all words from board using Trie
    public List<String> findWords(char[][] board, String[] words) {
        for (String w : words) insert(w);
        Set<String> found = new HashSet<>();
        for (int r = 0; r < board.length; r++)
            for (int c = 0; c < board[0].length; c++)
                searchBoard(board, r, c, root, found);
        return new ArrayList<>(found);
    }

    private void searchBoard(char[][] board, int r, int c, TrieNode node, Set<String> found) {
        if (r < 0 || r >= board.length || c < 0 || c >= board[0].length) return;
        char ch = board[r][c];
        if (ch == '#' || !node.children.containsKey(ch)) return;
        node = node.children.get(ch);
        if (node.isEnd) found.add(node.word);
        board[r][c] = '#'; // visited
        for (int[] d : new int[][]{{0,1},{0,-1},{1,0},{-1,0}})
            searchBoard(board, r+d[0], c+d[1], node, found);
        board[r][c] = ch;
    }

    static class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        boolean isEnd = false;
        String word = null;
    }

    public static void main(String[] args) {
        Trie trie = new Trie();
        trie.insert("apple"); trie.insert("app"); trie.insert("application");
        System.out.println(trie.search("app"));        // true
        System.out.println(trie.startsWith("appl"));   // true
        System.out.println(trie.autocomplete("app"));  // [app, apple, application]
    }
}
